src/sonarcloud/service.ts

```diff
@@ -0,0 +1,80 @@
+import fetch from "node-fetch"
+import { Config } from "../config"
+import { SonarCloudTokenCliProvider, SonarCloudTokenProvider } from "./token"
+import { SonarCloudProjectMetrics } from "./types"
+
+interface SonarCloudServiceProps {
+  config: Config
+  tokenProvider: SonarCloudTokenProvider
+}
+
+export class SonarCloudService {
+  private config: Config
+  private tokenProvider: SonarCloudTokenProvider
+
+  public constructor(props: SonarCloudServiceProps) {
+    this.config = props.config
+    this.tokenProvider = props.tokenProvider
+  }
+
+  public async getMetricsByProjectId(
+    SonarCloudOrganizationId: string,
+    SonarCloudProjectId: string,
+  ): Promise<SonarCloudProjectMetrics> {
+    const token = await this.tokenProvider.getToken()
+    if (token === undefined) {
+      throw new Error("Missing token for SonarCloud")
+    }
+
+    const response = await fetch(
+      `https://sonarcloud.io/api/measures/component?component=${encodeURIComponent(
+        SonarCloudOrganizationId,
+      )}_${encodeURIComponent(SonarCloudProjectId)}&metricKeys=coverage`,
+      {
+        method: "GET",
+        headers: {
+          Accept: "application/json",
+          Authorization: `token ${token}`,
+        },
+        agent: this.config.agent,
+      },
+    )
+
+    if (response.status === 401) {
+      process.stderr.write("Unauthorized - removing token\n")
+      await this.tokenProvider.markInvalid()
+    }
+    if (!response.ok && !(response.status === 401)) {
+      throw new Error(
+        `Response from SonarCloud not OK (${response.status}): ${JSON.stringify(
+          response,
+        )}`,
+      )
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    const metricsJson = (await response.json()) as {
```

Comment:
We may want to add all available fields in the response here (e.g., `id`) for future use-cases, and store the type separately in `types.ts` so it more easily can be reused